stablehlo/dialect/TypeInference.cpp

```diff
@@ -1330,17 +1330,31 @@ LogicalResult inferAllToAllOp(
 
   // If operand is ranked, size of split dimension should be a multiple of split
   // count.
-  auto splitDimSize = operandRankedType.getDimSize(splitDimension);
-  if (splitDimSize % splitCount != 0)
-    return emitOptionalError(
-        location, "split dimension has size ", splitDimSize,
-        ", expected to be a multiple of split_count ", splitCount);
   SmallVector<int64_t> resultShape(operandRankedType.getShape().begin(),
                                    operandRankedType.getShape().end());
-  resultShape[splitDimension] /= splitCount;
-  resultShape[concatDimension] *= splitCount;
-  inferredReturnShapes.emplace_back(resultShape,
-                                    operandRankedType.getElementType());
+  if (!isDynamicDimSize(resultShape[splitDimension]) &&
+      resultShape[splitDimension] % splitCount != 0) {
+    return emitOptionalError(
+        location, "split dimension has size ", resultShape[splitDimension],
+        ", expected to be a multiple of split_count ", splitCount);
+  }
+  resultShape[splitDimension] /=
+      isDynamicDimSize(resultShape[splitDimension]) ? 1 : splitCount;
+  resultShape[concatDimension] *=
+      isDynamicDimSize(resultShape[concatDimension]) ? 1 : splitCount;
```

Comment:
Please rebase now that #922 is merged. This will pull in the logic that already implements this part.